### PR TITLE
[fix]: [CDS-35484]: update corrupted instance stats migration

### PR DESCRIPTION
**Pull request by tchaurasiya**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32060)

### CODE REVIEW

File: 390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  private final String DEBUG_LINE = "UPDATE_CORRUPTED_INSTANCE_STATS_MIGRATION: ";
+
+  @Override
+  public void migrate() {
+    log.info("Running UpdateCorruptedInstanceStatsMigration");
+    Query<InfrastructureMapping> accountIdsQuery = wingsPersistence.createQuery(InfrastructureMapping.class)
+                                                       .filter("deploymentType", "HELM")
+                                                       .project("accountId", true)
+                                                       .project("_id", true);
+    Map<String, Set<String>> affectedAccounts = new HashMap<>();
```
Comment (shubham-in10se) [inline 390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java:60]: we are not using this mapping anywhere. We can just collect the `Set` of affected accountIds.

**Review by shubham-in10se — APPROVED** [review]

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 60 at r1](https://reviewable.io/reviews/harness/harness-core/32060#-MzuQfTZZflBZSWDJLma-r1-60:-MzxRe4x6CJnXL2NHoma:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/439d32fc8eb130fc887e0acbbf1544e59ba06d83/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L60)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

we are not using this mapping anywhere. We can just collect the `Set` of affected accountIds.
</blockquote></details>

Done.

**Review by shubham-in10se — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32060)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 63 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzxVtmQ3gyNFrxBiW-P:-MzxVtmRBVsrqlBMwZ6x:b6ad1bz) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L63)):*
> ```Java
>       }
>     } catch (Exception ex) {
>       log.error(StringUtils.join(DEBUG_LINE, "Error fetching affected accounts for migration"));
> ```

log the caught exception as well. It'll help in debugging the issue.
___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 104 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzxW0lt5bd6YPPONyZE:-MzxW0lt5bd6YPPONyZF:b-4sp618) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L104)):*
> ```Java
>       } catch (Exception ex) {
>         log.error(StringUtils.join(
>             DEBUG_LINE, format("Error updating corrupted instance stats for accountId %s ", accountId)));
> ```

always log the exception in error logs.
___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 124 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzxW8v0Doi8BBf_rgLS:-MzxW8v0Doi8BBf_rgLT:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L124)):*
> ```Java
>     } catch (Exception ex) {
>       log.error(StringUtils.join(
>           DEBUG_LINE, format("Error getting unique instance count for accountId %s, appId %s ", accountId, appId)));
> ```

same as above

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 63 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzxVtmQ3gyNFrxBiW-P:-N-2eeqA2AcanJt5Y67z:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L63)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

log the caught exception as well. It'll help in debugging the issue.
</blockquote></details>

Done.

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 104 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzxW0lt5bd6YPPONyZE:-N-2egNXEeiEo2knjbFk:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L104)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

always log the exception in error logs.
</blockquote></details>

Done.

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 124 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzxW8v0Doi8BBf_rgLS:-N-2ehSN1IClJeoXBUFY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L124)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32060)* status: 1 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 136 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzyCUmO2pfJhKFi2AZ3:-MzyCUmO2pfJhKFi2AZ4:b-kqabwl) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L136)):*
> ```Java
>                                                           .filter("isDeleted", false)
>                                                           .filter("containerInstanceKey", null)
>                                                           .field("deletedAt")
> ```

is query right? The logic should isDeleted == false || deletedAt <= toTime, in this case wouldn't it be  isDeleted == false && deletedAt <= toTime that always be false
___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 145 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzyBkkFE2xdQDVCRd5u:-MzyBkkG770bZVbU_JwQ:b6721cg) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L145)):*
> ```Java
>         .createAggregation(Instance.class)
>         .match(nonContainerFetchInstancesQuery)
>         .group(id(grouping("containerId", "containerInstanceKey.containerId"),
> ```

If containerInstanceKey is null here, how exactly this will work? 
Do we even need in this case aggregation? find() and count() wouldn't be enough? 
___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 160 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzyCFr_D8pQcsrPRNlN:-MzyCFr_D8pQcsrPRNlO:b-kqabwl) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L160)):*
> ```Java
>                                                        .filter("isDeleted", false)
>                                                        .filter("containerInstanceKey !=", null)
>                                                        .field("deletedAt")
> ```

is query right? The logic should isDeleted == false || deletedAt <= toTime, in this case wouldn't it be  isDeleted == false && deletedAt <= toTime that always be false
___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 54 at r3](https://reviewable.io/reviews/harness/harness-core/32060#-N-2qUDM4hnP5QPAC5e5:-N-2qUDM4hnP5QPAC5e6:b40k9tc) ([raw file](https://github.com/harness/harness-core/blob/11165b83bec66d2a2a39de29dfaef430904a6f18/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L54)):*
> ```Java
> 
>   // time of 21.02.2022 00:00:00 IST
>   private static final Instant FROM = Instant.ofEpochMilli(1645381800000L);
> ```

Why we're doing from 21st of feb? Shouldn't it be from start of march? 
___
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 57 at r3](https://reviewable.io/reviews/harness/harness-core/32060#-N-2qjZP-W6_cyZPdmsX:-N-2qjZP-W6_cyZPdmsY:b-1vdjyr) ([raw file](https://github.com/harness/harness-core/blob/11165b83bec66d2a2a39de29dfaef430904a6f18/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L57)):*
> ```Java
> 
>   // time of 08.03.2022 00:00:00 GMT
>   private static final long CREATED_AT_FROM_TIMESTAMP = 1646697600000L;
> ```

Let's give some buffer here of 2 days (from 06.03 to 25.03), just to be sure
___
*[400-rest/src/main/java/io/harness/mongo/index/migrator/ContainerIdAggregate.java, line 9 at r3](https://reviewable.io/reviews/harness/harness-core/32060#-N-2rFHBBYBVdh4nanR_:-N-2rFHBBYBVdh4nanRa:bjpefxk) ([raw file](https://github.com/harness/harness-core/blob/11165b83bec66d2a2a39de29dfaef430904a6f18/400-rest/src/main/java/io/harness/mongo/index/migrator/ContainerIdAggregate.java#L9)):*
> ```Java
> @Data
> @Entity(noClassnameStored = true)
> public class ContainerIdAggregate {
> ```

Is this required anymore? as well as ContainerIdAndNamespace

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 54 at r3](https://reviewable.io/reviews/harness/harness-core/32060#-N-2qUDM4hnP5QPAC5e5:-N-8D9nK5_4twkAe_Im8:b-urdqbd) ([raw file](https://github.com/harness/harness-core/blob/11165b83bec66d2a2a39de29dfaef430904a6f18/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L54)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Why we're doing from 21st of feb? Shouldn't it be from start of march? 
</blockquote></details>

Done. using 01.03.2022 now

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 57 at r3](https://reviewable.io/reviews/harness/harness-core/32060#-N-2qjZP-W6_cyZPdmsX:-N-8DIJh5MSnqz7VbM5h:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/11165b83bec66d2a2a39de29dfaef430904a6f18/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L57)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Let's give some buffer here of 2 days (from 06.03 to 25.03), just to be sure
</blockquote></details>

Done.

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 145 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzyBkkFE2xdQDVCRd5u:-N-8DLG1ETQV7l0Xymg8:bxr9kb6) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L145)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

If containerInstanceKey is null here, how exactly this will work? 
Do we even need in this case aggregation? find() and count() wouldn't be enough? 
</blockquote></details>

Done.With new logic no need of containerInstanceKey

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 136 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzyCUmO2pfJhKFi2AZ3:-N-8DPqR6SeVqZQ97edr:b-8erf27) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L136)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

is query right? The logic should isDeleted == false || deletedAt <= toTime, in this case wouldn't it be  isDeleted == false && deletedAt <= toTime that always be false
</blockquote></details>

Done.Fixed.

**Comment by tchaurasiya** [comment]
*[390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java, line 160 at r2](https://reviewable.io/reviews/harness/harness-core/32060#-MzyCFr_D8pQcsrPRNlN:-N-8DSNR-mqNlbjWbAVO:b-8erf27) ([raw file](https://github.com/harness/harness-core/blob/c8f24a6ec6a3ef97ffa27dbaa9abe4efd736a6bb/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L160)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

is query right? The logic should isDeleted == false || deletedAt <= toTime, in this case wouldn't it be  isDeleted == false && deletedAt <= toTime that always be false
</blockquote></details>

Done.Fixed.

**Comment by tchaurasiya** [comment]
*[400-rest/src/main/java/io/harness/mongo/index/migrator/ContainerIdAggregate.java, line 9 at r3](https://reviewable.io/reviews/harness/harness-core/32060#-N-2rFHBBYBVdh4nanR_:-N-8DV8zBIP2V5XcH6xV:b-w3f6qd) ([raw file](https://github.com/harness/harness-core/blob/11165b83bec66d2a2a39de29dfaef430904a6f18/400-rest/src/main/java/io/harness/mongo/index/migrator/ContainerIdAggregate.java#L9)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Is this required anymore? as well as ContainerIdAndNamespace
</blockquote></details>

Done.Removed unused classes.

**Review by abosii-harness — APPROVED** [review]
> Reviewed 2 of 3 files at r4, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32060)* status: 1 change requests. 3. unresolved discussions. 

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java` line 252 at r6](https://reviewable.io/reviews/harness/harness-core/32060#-N0enq60BT66FZG4Rzpb:-N0enq60BT66FZG4Rzpc:brx6b96) ([raw file](https://github.com/harness/harness-core/blob/afca5554f4fd05f8828e7144949188665bf7ce45/390-db-migration/src/main/java/io/harness/migrations/all/UpdateCorruptedInstanceStatsMigration.java#L252)):*
> ```Java
>   }
> 
>   private Map<Timestamp, Map<String, TimeSeriesBatchEventInfo.DataPoint>> processResultSetToCreateDataPoints(
> ```

nit: is this method used anywhere?

**Comment by harness-david** [comment]
As a sanity check these are the numbers of TimeScaleDB records that will be backed up

prod-1
INSTANCE_STATS_COUNT = 37,770,770
INSTANCE_STATS_HOUR_COUNT = 1,919,836
INSTANCE_STATS_DAY_COUNT = 78,960

prod-2
INSTANCE_STATS_COUNT = 0
INSTANCE_STATS_HOUR_COUNT = 0
INSTANCE_STATS_DAY_COUNT = 0

prod-3
INSTANCE_STATS_COUNT = 1,535,473  (only for CITIPROD accountid 6pitvBmtSMKNZU3gANq01Q)
INSTANCE_STATS_HOUR_COUNT = 0
INSTANCE_STATS_DAY_COUNT = 0

**Pull request merged**